File: ruby/linked-list/linked_list.rb
class Deque
  class Node
    attr_reader :value
    attr_accessor :prev, :succ

    def initialize(value, prev: self, succ: self)
      @value = value
      @prev = prev
      @succ = succ
    end
  end
  private_constant :Node

  def initialize
    @head = nil
  end

  def push(value)
    if head
      tail = head.prev
      node = Node.new(value, prev: tail, succ: head)
      tail.succ = node
      head.prev = node
    else
      self.head = Node.new(value)
    end
  end

  def pop
    if (tail = head.prev) == head
      self.head = nil
    else
      head.prev = tail.prev
      tail.prev.succ = head
    end
    tail.value
  end

  def shift
    if (node = head) == head.succ
      self.head = nil
    else
      self.head = node.succ
      head.prev = node.prev
      head.prev.succ = head
    end
    node.value
  end

  def unshift(value)
    if head
      node = Node.new(value, prev: head.prev, succ: head)
      head.prev = node
      self.head = node
    else
      self.head = Node.new(value)
    end
  end

  private

  attr_accessor :head
end

File: typescript/linked-list/linked-list.ts
type MaybeNull<T> = T | null
type MaybeVoid<T> = T | void
type Links<T> = { prev: Node<T>, next: Node<T> } | {}
type HeadNode<T> = MaybeNull<Node<T>>

class Node<T> {
  readonly value: T
  prev: Node<T>
  next: Node<T>

  constructor(value: T, links: Links<T> = {}) {
    this.value = value
    this.prev = ("prev" in links) ? links.prev : this
    this.next = ("next" in links) ? links.next : this
  }
}

export default class LinkedList<T> {
  private head: HeadNode<T>

  constructor() {
    this.head = null
  }

  push(value: T): void {
    const head: HeadNode<T> = this.head

    if (!head) {
      this.head = new Node(value)
      return
    }

    const tail: Node<T> = head.prev
    const newTail: Node<T> = new Node(value, { prev: tail, next: head })
    tail.next = newTail
    head.prev = newTail
  }

  pop(): MaybeVoid<T> {
    const head: HeadNode<T> = this.head

    if (!head) {
      return
    }

    const tail: Node<T> = head.prev

    if (tail === head) {
      this.head = null
    } else {
      head.prev = tail.prev
      tail.prev.next = head
    }
    return tail.value
  }

  shift(): MaybeVoid<T> {
    const head: HeadNode<T> = this.head

    if (!head) {
      return
    }

    const newHead: Node<T> = head.next

    if (newHead === head) {
      this.head = null
    } else {
      newHead.prev = head.prev
      head.prev.next = newHead
      this.head = newHead
    }

    return head.value
  }

  unshift(value: T): void {
    const head: HeadNode<T> = this.head

    if (!head) {
      this.head = new Node(value)
      return
    }

    const newHead: Node<T> = new Node(value, { prev: head.prev, next: head })
    head.prev = newHead
    this.head = newHead
  }

  count(): number {
    const head: HeadNode<T> = this.head

    if (!head) {
      return 0
    }

    let count = 1
    let current: Node<T> = head

    while (current.next !== head) {
      count += 1
      current = current.next
    }

    return count
  }

  delete(value: T): void {
    const head: HeadNode<T> = this.head

    if (!head) {
      return
    }

    if (head.value === value) {
      this.shift()
      return
    }

    let current: Node<T> = head

    while (current.value !== value && current.next !== head) {
      current = current.next
    }

    if (current.value === value) {
      current.prev.next = current.next
      current.next.prev = current.prev
    }
  }
}
